### PR TITLE
[Dashboard] Remove unnecessary gRPC call in `dashboard/reporter_agent` on worker nodes

### PR DESCRIPTION
**Pull request by jon-chuang**
## Why are these changes needed?

`cluster_stats` is only used by head node agent.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jon-chuang** [comment]
@rkooo567 do you think an even better solution is to move this logic to reporter_head.py instead?

**Comment by rkooo567** [comment]
cc @wuisawesome 

If it is needed only for a head node, why don't we add it to reporter_head.py in the first place?

**Comment by jon-chuang** [comment]
@rkooo567 as I mentioned, this would neccessitate starting another prometheus client in the reporter_head, as the consumer is the prometheus server.

**Comment by wuisawesome** [comment]
I don't know the dashboard code at all (outside of the snapshot service). All I know is that I think `/api/cluster_status` (I don't know what it depensd on though).

File: dashboard/modules/reporter/reporter_agent.py [hunk 1]
```diff
@@ -589,9 +589,11 @@ async def _perform_iteration(self, publisher):
         """Get any changes to the log files and push updates to kv."""
         while True:
             try:
-                formatted_status_string = internal_kv._internal_kv_get(
-                    DEBUG_AUTOSCALING_STATUS
-                )
+                formatted_status_string = None
```
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_agent.py:592]: Don't we also need to put other logic related to formatted_status_string to this block? Also, please add a comment why it is only for a head node. 

```python3
stats = self._get_all_stats()
formatted_status_string = None
# Cluster stats don't need to be recorded on all agents, so we only report it in a head node. 
if self._is_head_node:
    formatted_status_string = internal_kv._internal_kv_get(DEBUG_AUTOSCALING_STATUS)
    cluster_stats = (
        json.loads(formatted_status_string.decode())
            if formatted_status_string else {})
    records_reported = self._record_stats(stats, cluster_stats)
    self._metrics_agent.record_reporter_stats(records_reported)
await publisher.publish_resource_usage(self._key, jsonify_asdict(stats))
```

Comment (jon-chuang) [reply dashboard/modules/reporter/reporter_agent.py:592]: No, the logic is quite convoluted. Basically, metrics agent will only log AUTOSCALING_STATUS on head node agent, but log other stats on non-head-node.

I think I will make it less confusing

**Comment by jon-chuang** [comment]
Appears all of the test failures have been due to https://github.com/ray-project/ray/commit/d27e73f85171ca23ed74175a3cdba6e555c4d540. Merging master again...

**Review by rkooo567 — APPROVED** [review]

File: dashboard/modules/reporter/reporter_agent.py [hunk 2]
```diff
@@ -678,24 +678,29 @@ def _record_stats(self, stats, cluster_stats):
...
+                        if formatted_status_string
+                        else {}
+                    )
+                else:
+                    cluster_stats = {}
 
                 stats = self._get_all_stats()
-                records_reported = self._record_stats(stats, cluster_stats)
-                self._metrics_agent.record_reporter_stats(records_reported)
+                records_to_report = self._get_records_to_report(stats, cluster_stats)
```
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_agent.py:702]: In this case, what about we just split `_get_records_to_report ` into two functions? 

```python3
if head:
    cluster_stats_to_report = self._get_records_to_report(stats, cluster_stats)
    self._metrics_agent.record_reporter_stats(cluster_stats_to_report)
node_stats_to_report = self._get_records_to_report(stats, cluster_stats)
self._metrics_agent.record_reporter_stats(node_stats_to_report)
```

Comment (jon-chuang) [reply dashboard/modules/reporter/reporter_agent.py:702]: Maybe I should just change cluster_stats to an optional parameter then

**Review by rkooo567 — COMMENTED** [review]
Can you add a test?

**Comment by jon-chuang** [comment]
does every PR need a test? I think the behaviour here is already covered by existing test?

**Comment by rkooo567** [comment]
Yeah if you change something, it is ideal to add tests to avoid the same type regression. Maybe we can just check the cluster status is not updated to other nodes than a head node?

**Comment by jon-chuang** [comment]
Added the neccesary test. 

The test checks that only the prometheus client on the head node will return cluster stats.

**Pull request closed without merging**